tuta-sdk/rust/sdk/src/blobs/blob_access_token_facade.rs

```diff
@@ -0,0 +1,147 @@
+use crate::blobs::blob_access_token_cache::BlobAccessTokenCache;
+use crate::crypto::randomizer_facade::RandomizerFacade;
+use crate::custom_id::CustomId;
+use crate::date::DateProvider;
+use crate::entities::storage::{BlobAccessTokenPostIn, BlobServerAccessInfo, BlobWriteData};
+use crate::generated_id::GeneratedId;
+#[cfg_attr(test, mockall_double::double)]
+use crate::services::service_executor::ResolvingServiceExecutor;
+use crate::services::storage::BlobAccessTokenService;
+use crate::services::ExtraServiceParams;
+use crate::tutanota_constants::ArchiveDataType;
+use crate::ApiCallError;
+use base64::prelude::BASE64_URL_SAFE_NO_PAD;
+use base64::Engine;
+use std::sync::Arc;
+
+/// The BlobAccessTokenFacade requests blobAccessTokens from the BlobAccessTokenService to get
+/// or post to the BlobService (binary blobs) or DefaultBlobElementResource (instances).
+/// All tokens are cached.
+pub(crate) struct BlobAccessTokenFacade {
+	cache: BlobAccessTokenCache,
+	randomizer_facade: RandomizerFacade,
+	service_executor: Arc<ResolvingServiceExecutor>,
+}
+
+#[cfg_attr(test, mockall::automock)]
+impl BlobAccessTokenFacade {
+	pub fn new(
+		randomizer_facade: RandomizerFacade,
+		service_executor: Arc<ResolvingServiceExecutor>,
+		date_provider: Arc<dyn DateProvider>,
+	) -> Self {
+		Self {
+			cache: BlobAccessTokenCache::new(date_provider),
+			randomizer_facade,
+			service_executor,
+		}
+	}
+
+	/// Requests a token that allows uploading blobs for the given ArchiveDataType and ownerGroup.
+	pub async fn request_write_token(
+		&self,
+		archive_data_type: ArchiveDataType,
+		owner_group_id: &GeneratedId,
+	) -> Result<BlobServerAccessInfo, ApiCallError> {
+		let archive_data_type_discriminant = archive_data_type.discriminant();
+		let owner_group_id_clone = owner_group_id.clone();
+		let loader = move || async move {
+			let post_in: BlobAccessTokenPostIn = BlobAccessTokenPostIn {
+				_format: 0,
+				archiveDataType: Some(archive_data_type_discriminant),
+				read: None,
+				write: Some(BlobWriteData {
+					_id: Some(CustomId(
+						BASE64_URL_SAFE_NO_PAD
+							.encode(self.randomizer_facade.generate_random_array::<4>()),
+					)),
+					archiveOwnerGroup: owner_group_id_clone,
+				}),
+			};
+			self.service_executor
+				.post::<BlobAccessTokenService>(post_in, ExtraServiceParams::default())
+				.await
+				.map(|r| r.blobAccessInfo)
+		};
+
+		self.cache
+			.try_get_token(
+				&make_write_cache_key(owner_group_id, archive_data_type),
+				loader,
+			)
+			.await
+	}
+
+	/// Remove a given write token from the cache.
+	pub fn evict_write_token(
+		&self,
+		archive_data_type: ArchiveDataType,
+		owner_group_id: &GeneratedId,
+	) {
+		let key = make_write_cache_key(owner_group_id, archive_data_type);
+		self.cache.evict(&key);
+	}
+}
+
+pub(crate) fn make_write_cache_key(
```

Comment:
should we make some struct that is a key? both id and type are tiny and we could store them directly on the map if the key was struct.

A separate type might also be more type-safe

Reply:
good idea